### DISCUSSION TITLE
Some word segmentation results are different than we get in ICU4C

### DISCUSSION THREAD

**Discussion by riajain0412** [unicode-org/icu4x#3522] [General]
I was comparing results of text segmentation between ICU4C and ICU4X for SEA languages but I found some disparity between the results. Listing down the few strings which are having different result in ICU4X and ICU4C.
1. Khmer string មនុស្សទាំងអស់ is giving 13 index as a breakpoint in ICU4X while ICU4C gives 6
2. ຮ່ສົ່ສີ 5 in ICU4C while 7 in ICU4X 
3. กระเพรา 3 in ICU4C while 7 in ICU4X
and many other strings
So, I wanted to confirm that are these expected results?

I'm using the full data blob with all keys and locales.

**Comment by sffc** [comment]
Which LSTM constructors are you using? Please verify that you can reproduce these results with the dictionary constructors.

**Reply by riajain0412** [reply]
I'm doing this with dictionary constructors.

**Reply by sffc** [reply]
OK, dictionary behavior is intended to be the same between ICU4C and ICU4X. @makotokato or @aethanyc can you reproduce?

**Reply by aethanyc** [reply]
I didn't reproduce on ICU4C. Here my test result for the word break points:

Testcase 1 Khmer string `មនុស្សទាំងអស់` , I got `[0, 6, 13]`.
Testcase 2 Lao string `ຮ່ສົ່ສີ`, I got `[0, 3, 4, 6, 7]`.
Testcase 3 Thai string `กระเพรา`, I got `[0, 3, 7]`. 

I reproduce the results by using `WordSegmenter::try_new_dictionary_with_buffer_provider` and replacing `s` in https://github.com/unicode-org/icu4x/blob/a8ef6735ca67554463ba8d9a58a755c71fd55fef/components/segmenter/tests/complex_word.rs#L57-L71

Then running `cargo test --all-features word_break_my` under `components/segmenter` in icu4x repo.

If index 13 in testcase 1 and index 7 in testcase 2 are the only word break points in the results (other than index 0), I guess the dictionary data for Khmer or Lao is probably not loaded.

**Comment by riajain0412** [comment]
I'm loading full data blob in my C++ code. How to confirm that whether dictionaries are loaded or not? 

And also which all keys are needed for word segmenter? I was trying to create a data blob for dictionary based word segmenter for SEA language only. I'm including only segmenter/word@1 and segmenter/dictionary/wl_ext@1.

**Reply by aethanyc** [reply]
I've tried the following code, and I need these keys `segmenter/word@1` , `segmenter/grapheme@1`,  `segmenter/dictionary/w_auto@1` , and `segmenter/dictionary/wl_ext@1` for the code to not panic.

```Rust
    let blob = std::fs::read("my_data_blob.postcard").expect("Failed to read file");
    let buffer_provider = BlobDataProvider::try_new_from_blob(blob.into_boxed_slice())
        .expect("Failed to initialize Data Provider");
    let segmenter = WordSegmenter::try_new_dictionary_with_buffer_provider(&buffer_provider)
        .expect("Data exists");
```

The dictionary & lstm loading setup has been changed since I last looked. 

@sffc @robertbastian Do we support loading partial dictionaries such as loading only SEA languages but not Japanese/Chinese?

**Reply by sffc** [reply]
You can apply locale filtering as normal in datagen to select the choice of dictionaries.

**Reply by aethanyc** [reply]
Am I reading code correctly that we just bail out from `try_new_dict` if grapheme or one of the dictionaries failed to load?
https://github.com/unicode-org/icu4x/blob/02cc1d44116c3143526ea707a2b8ffdf2ebc2a0c/components/segmenter/src/complex/mod.rs#L95-L119

**Reply by sffc** [reply]
It works; the `try_load` function returns a `Result<Option<>>` which resolves to `Ok(None)` if the provider returned a `MissingLocale` error.

**Reply by aethanyc** [reply]
@sffc Thanks for the answer!

> You can apply locale filtering as normal in datagen to select the choice of dictionaries.

That means if you don't want Japanese/Chinese dictionary, don't include `ja` in the ```--locales```  for the `icu4x-datagen`. For example, try  

```
icu4x-datagen --keys segmenter/word@1 segmenter/grapheme@1 segmenter/dictionary/w_auto@1 segmenter/dictionary/wl_ext@1 --locales my km lo th
```

> How to confirm that whether dictionaries are loaded or not?

On ICU4X `main` branch, you can enable `log_error_context` Cargo feature for the `icu_provider` crate in your project's `Cargo.toml`. Then use https://crates.io/crates/env_logger to dump log.

When you iterate a `WordBreakIterator` with a Khmer string but without the Khmer dictionary loaded, it will print something like 

```
[2023-06-14T05:03:31Z WARN  icu_provider::error] ICU4X data error: No segmentation model for language: km
```

The related code that prints the warning is at
https://github.com/unicode-org/icu4x/blob/02cc1d44116c3143526ea707a2b8ffdf2ebc2a0c/components/segmenter/src/complex/mod.rs#L43-L68

**Reply by riajain0412** [reply]
I'm doing this in C++. So how can logs be generated in C++?

**Reply by riajain0412** [reply]
> ```
> icu4x-datagen --keys segmenter/word@1 segmenter/grapheme@1 segmenter/dictionary/w_auto@1 segmenter/dictionary/wl_ext@1 --locales my km lo th
> ```
>
I was using data in blob format so used this command to generate data blob and then tried to load that blob in my C++ code but it's not giving expected segmenter results.

**Reply by riajain0412** [reply]
Below is the C++ code which I'm running. 
```cpp
int main(){
    cout<<"Press enter to start"<<endl;
    cin.get();
    UnicodeString str("ກຳລັງກວດສອບຈຸດຢຸດ");
    printUnicodeString(str);
    cout<<endl;
    std::ifstream fin("/Users/riajain/Desktop/ria/sample_client/my_data_blob", std::ifstream::ate | std::ifstream::binary);

    fin.seekg(0, std::ios::end);
    std::streamsize blob_size = fin.tellg();
    cout<<"size buffer "<<blob_size<<endl;
    fin.seekg(0, std::ios::beg);

    char* buffer = new char[blob_size];
    fin.read(buffer, blob_size);
    cout<<buffer<<endl;

    std::vector<uint8_t> f_buffer;
    f_buffer.insert(f_buffer.end(), reinterpret_cast<const uint8_t*>(buffer), reinterpret_cast<const uint8_t*>(buffer+ blob_size));
    const diplomat::span<const uint8_t> blob(f_buffer.data(), f_buffer.size());
    const uint16_t* u16Ptr = reinterpret_cast<const uint16_t*>(str.getBuffer());
    
    const auto provider_result = ICU4XDataProvider::create_from_byte_slice(blob).ok().value();
    const auto segmenter_auto = ICU4XWordSegmenter::create_auto(provider_result).ok().value();
    const ICU4XWordSegmenter* segmenters[] = {&segmenter_auto};
    
    size_t sizeInBytes = str.length() * sizeof(UChar);
    const size_t size = sizeInBytes / sizeof(uint16_t); // Compute the number of elements in the array
    diplomat::span<const uint16_t> span(u16Ptr,size);
    auto iterator = segmenter_auto.segment_utf16(span);

    int32_t breakpoint = iterator.next();
    int32_t breakpoint2 = iterator.next();
    cout<<breakpoint<<endl;
    cout<<breakpoint2<<endl;
    delete[] buffer;
}
```

**Reply by riajain0412** [reply]
And also can we create .a file for data files?

**Reply by robertbastian** [reply]
You can use `xxd` to convert any binary file into a C header that defines an `unsigned char[]`, and then compile that into a static library: 

```
xxd -i -C data.postcard > data.h
clang data.h -o data.o
ar r data.a data.o
```

The resulting archive is 3x bigger than the original file, so I'm probably missing some arguments to strip debug info.

**Reply by robertbastian** [reply]
For logging in C++ you want to enable the `logging` and `simple_logger` features when you build `icu_capi`. These are included in `cpp_default`, so you can do `cargo build -p icu_capi_staticlib --features cpp_default`

**Reply by riajain0412** [reply]
Okay, I have created icu_capi_staticlib using the above cargo command only but where does log get generated?

**Reply by robertbastian** [reply]
It logs to stdout.

**Reply by Manishearth** [reply]
You d have to initialize ICU4XLogger in c++ first

**Reply by robertbastian** [reply]
Ah of course, https://unicode-org.github.io/icu4x/docs/ffi/cpp/logging_ffi.html

**Reply by riajain0412** [reply]
Okay, I'll try this out. Thank you.

**Reply by riajain0412** [reply]
@robertbastian , is cargo command the only way to create icu_capi_staticlib?

**Reply by robertbastian** [reply]
Cargo is the Rust package manager. You don't *need* it, you can call `rustc` directly, but then you either need some other package manager, or compile every dependency manually.

**Reply by riajain0412** [reply]
Ohh Okay. Thank you for your help.
